Migration guide 3.0.0 (#977)
neo4j/graphql: docs/modules/ROOT/pages/guides/v3-migration/index.adoc

[[v3-migration]]
= 3.0.0 Migration
This document lists all breaking changes from version 2.x.y to 3.0.0 and how to update.

== How to upgrade
Simply update `@neo4j/graphql` using npm or your package manager of choice:

[source, bash, indent=0]
----
npm update @neo4j/graphql
----

== Async schema generation
Schema generation is now asynchronous. Instead of using the property `schema`, now the method `getSchema` will return the schema
as a _Promise_. This means that creating a server now requires awaiting for that method:

Instead of
[source, JavaScript, indent=0]
----
const neoSchema = new Neo4jGraphQL({ typeDefs, driver });
const server = new ApolloServer({
    schema: neoSchema.schema,
});
----

Now you'll need to do the following:

[source, JavaScript, indent=0]
----
const neoSchema = new Neo4jGraphQL({ typeDefs, driver });
const schema = await neoSchema.getSchema();
const server = new ApolloServer({

>     schema: schema

Reviewer: nitpicking: couldn't this just be `schema`?

Author: Yes :sweat_smile: , I decided to go with this as some people may not be completely familiar with object destructuring, but happy to change if you think it is cleaner that way

Reviewer: Yes I see. Good point! Let's opt for the most familiar version and leave it as is.

});

> ----

Reviewer 2: Possibly also worth documenting this using `.then`? I imagine this will be appropriate for a lot of users.

```
----
const neoSchema = new Neo4jGraphQL({ typeDefs, driver });
neoSchema.getSchema().then((schema) => {
  const server = new ApolloServer({
    schema: schema
  });
});
----
```

Author: mmm, documenting both would be heavily redundant, so I guess it is promises vs async await

Thoughts?

Reviewer: Does this come down to a matter a of (personal) preference? Then I vote for async await due to its slightly better readability IMO. But if the majority of our users are more familiar to promises then we should opt for that.

Author: I agree with you, although I think most of our examples are already riddled with async/await. Thoughts on this @darrellwarde @danstarns ?

== Relationship changes
This release contains an overhaul of our relationship validations, this will require a few changes to the schema.

=== Many to * relationships
To improve consistency and validation, **all** "many to *" relationships need to be defined as _required_ in the schema:

[source, graphql, indent=0]
----
type Movie {

>     actors: [Actor!]! @relationship(type: "ACTED_IN", direction: IN)

Reviewer 3: What will be the API behaviour if Movie does not have any relationships to Actors? I'm hoping this change doesn't mean there'll be an error.

Reviewer 2: Behaviour will be identical - this always returned an empty List, and this new type definition constraint matches the possible return values.

    director: Director @relationship(type: "DIRECTED", direction: IN)
}
----

Note that any other notation, such as `[Actor]` or `[Actor!]` will **not** be valid. "One to one" relationships

> such as `director` remain unchanged.

Reviewer: .. such as `Director` remain ..

> === Relationship cardinality

Reviewer 3: Are there any more details available of the effect this will have? e.g. will the server start to error if a rogue relationship is created in the DB? Can this functionality be turned off?

Author: Hi @wheresrhys, we will try to be as comprehensive with potential pitfalls as possible. As you mention, there is a change the server may return an error on **mutations**  if your database has rogue relationships. Please keep in mind that if that is the case, the current behaviour is already undetermined and **will** cause problems, at least for now we do not plan on making this optional, as the current behaviour is actually buggy.

Runtime checks for relationships number of connections have been rolled in with this release, this means that some
databases with inconsistent relationships between the schema definition and the actual data may now fail in some queries.
For those cases, please ensure that the database is following your schema definition or update the schema to reflect the
actual existing relationships.

== Count query no longer supported
Queries using `count` at the root level are no longer supported. For example:
[source, graphql, indent=0]
----
Query {
    usersCount
}
----

The same operation, can now be achieved with a xref::queries#_counting_using_aggregation[count aggregation] query:

[source, graphql, indent=0]
----
Query {
    usersAggregate {
        count
    }
}
----

== Relationship filters
`where` filters for relationship queries now explicitly state `ALL`, `NONE`, `SINGLE`, and `SOME` as part of filter name.

> Queries using old relationship filters, will now need to use `{relationship}_NONE`. For example:

Reviewer 2: ```suggestion
Queries using old relationship filters, will now need to use `{relationship}_SOME`. For example:
```

[source, graphql, indent=0]
----
Query {
  movies(where: {
    actors: {
      name: "John"
    }
  }) {
    title
  }
}
----

Should be:

[source, graphql, indent=0]
----
Query {
  movies(where: {
    actors_SOME: {
      name: "John"
    }
  }) {
    title
  }
}
----

And, instead of `_NOT`, `_NONE` should be used.

NOTE: Old queries will still work in this release, but are marked as `@deprecated` and will not be available in the future.

== `@ignore` directive renamed to `@computed`
To better reflect its intended usage, the `@ignore` directive is now named `@computed`, behaviour is unchanged, so you just need to
rename this directive in your schema.

== Auth plugin system

> https://auth0.com/docs/secure/tokens/json-web-tokens/json-web-key-sets[JSON Web Key Sets] support through `jwksEndpoint` option is no longer

Reviewer 2: This is more than just `jwksEndpoint`? I would suggest this should have two subsections, one for JWT users and one for JWKS users.

Author: Agree, I think I'll wait until auth is merged to fix this, as we will probably want to link to the updated docs to avoid duplication

provided directly by `@neo4j/graphql`.

This feature is now available as a plugin `@neo4j/graphql-plugin-auth` that can be installed if needed:

[source, JavaScript, indent=0]
----
const { JWTPlugin } = require("@neo4j/graphql-plugin-auth");

const neoSchema = new Neo4jGraphQL({
     typeDefs,
     plugins: {
         jwt: new JWTPlugin({
             secret: "secret",
         }),
     },
 });
----

== Types plurals changes
To improve consistency, some automatically generated plurals (e.g. `createActors`) have changed. This may cause issues if
your types use conventions such as `snake_case`.

Because of this, you may find generated queries and mutations may have different names. If you encounter this problem,

> please update your clients to use the new query names or use the `plural` option in the xref::type-definitions/database-mapping.adoc#_plural[`@node directive`]

Reviewer: More of an accessibility note. The link to `@node directive` is very helpful but it wasn't very obvious to me that this is a link.

Author: What alternative could we use? just "@node directive" without code blocks?

Reviewer: The alternative without code blocks seems good. But on the other hand this may also just be some useless nitpicking from my side 😅 . We can leave it as is.

Author: If you think it helps accessibility I'm happy to change it
```suggestion
please update your clients to use the new query names or use the `plural` option in the xref::type-definitions/database-mapping.adoc#_plural[@node directive]
```

to force a custom plural value.

== Types changes
Some automatically generated types have changed to improve consistency.
These should not require any changes from most developers, unless types names are directly used.

If you are using auto-generated type names, please check the full list of xref::guides/v3-migration/generated-types.adoc[generated types changes].

== Neo4j support

> Neo4j 4.1 is no longer supported in 3.0

Reviewer 2: ```suggestion
Neo4j 4.1 is no longer supported in 3.0.0, inline with the [supported versions list](https://neo4j.com/developer/kb/neo4j-supported-versions/).
```

== GraphQL support

> GraphQL 15 is no longer supported, please migrate to GraphQL 16

Reviewer 2: A bit more boilerplate, but I prefer something like this:

```suggestion
`graphql@^15.0.0` is no longer supported, please upgrade to `graphql@^16.0.0` using `npm` or the package manager of your choice.
```

[[v3-migration]]
= 3.0.0 Migration
This document lists all breaking changes from version 2.x.y to 3.0.0 and how to update.

== How to upgrade
Simply update `@neo4j/graphql` using npm or your package manager of choice:

[source, bash, indent=0]
----
npm update @neo4j/graphql
----

> == Async schema generation

Reviewer 2: ```suggestion
== Asynchronous schema generation
```

Schema generation is now asynchronous. Instead of using the property `schema`, now the method `getSchema` will return the schema

> as a _Promise_. This means that creating a server now requires awaiting for that method:

Reviewer 2: ```suggestion
as a `Promise`. This means that creating a server now requires awaiting for that method:
```

Instead of
[source, JavaScript, indent=0]
----
const neoSchema = new Neo4jGraphQL({ typeDefs, driver });
const server = new ApolloServer({
    schema: neoSchema.schema,
});
----

Now you'll need to do the following:

[source, JavaScript, indent=0]
----
const neoSchema = new Neo4jGraphQL({ typeDefs, driver });
neoSchema.getSchema().then((schema) => {
    const server = new ApolloServer({
        schema: schema
    });
});
----

== Relationship changes

> This release contains an overhaul of our relationship validations, this will require a few changes to the schema.

Reviewer 2: ```suggestion
This release contains an overhaul of our relationship validations, which will require a few changes to the schema.
```

> === Many to * relationships

Reviewer 2: ```suggestion
=== Many-to-* relationships
```

> To improve consistency and validation, **all** "many to *" relationships need to be defined as _required_ in the schema:

Reviewer 2: ```suggestion
To improve consistency and validation, **all** "many-to-*" relationships need to be defined as _required_ in the schema:
```

[source, graphql, indent=0]
----
type Movie {
    actors: [Actor!]! @relationship(type: "ACTED_IN", direction: IN)
    director: Director @relationship(type: "DIRECTED", direction: IN)
}
----

> Note that any other notation, such as `[Actor]` or `[Actor!]` will **not** be valid. "One to one" relationships

Reviewer 2: ```suggestion
Note that any other notation, such as `[Actor]` or `[Actor!]` will **not** be valid. "One-to-one" relationships
```

such as `Director` remain unchanged.

=== Relationship cardinality

> Runtime checks for relationships number of connections have been rolled in with this release, this means that some

Reviewer 2: ```suggestion
Runtime checks for "one-to*" relationships have been added in this release, ensuring that the correct number of relationships exist. This means that some
```

> databases with inconsistent relationships between the schema definition and the actual data may now fail in some queries.

Reviewer 2: Why would this have happened? It will be worthwhile giving some background as to why some databases might have ended up in this state.

Author: I can think of manual changes to the DB, bugs, or changes in typedefs as the main culprits, anything else we should list there?

Reviewer 2: More like, due to the fact that we never enforced this before, it's likely that through normal use of our library before this version, your database may have ended up in this state.

For those cases, please ensure that the database is following your schema definition or update the schema to reflect the
actual existing relationships.

== Count query no longer supported
Queries using `count` at the root level are no longer supported. For example:
[source, graphql, indent=0]
----

> Query {

Reviewer 2: ```suggestion
query {
```

    usersCount
}
----

The same operation, can now be achieved with a xref::queries#_counting_using_aggregation[count aggregation] query:

[source, graphql, indent=0]
----

> Query {

Reviewer 2: ```suggestion
query {
```

    usersAggregate {
        count
    }
}
----

== Relationship filters
`where` filters for relationship queries now explicitly state `ALL`, `NONE`, `SINGLE`, and `SOME` as part of filter name.

Queries using old relationship filters, will now need to use `{relationship}_SOME`. For example:

[source, graphql, indent=0]
----

> Query {

Reviewer 2: ```suggestion
query {
```

  movies(where: {
    actors: {
      name: "John"
    }
  }) {
    title
  }
}
----

Should be:

[source, graphql, indent=0]
----

> Query {

Reviewer 2: ```suggestion
query {
```

  movies(where: {
    actors_SOME: {
      name: "John"
    }
  }) {
    title
  }
}
----

And, instead of `_NOT`, `_NONE` should be used.

> NOTE: Old queries will still work in this release, but are marked as `@deprecated` and will not be available in the future.

Author: mmm, I wasn't completely sure, but yeah, makes sense

== `@ignore` directive renamed to `@computed`
To better reflect its intended usage, the `@ignore` directive is now named `@computed`, behaviour is unchanged, so you just need to
rename this directive in your schema.

== Auth plugin system
Auth setup now relies on _plugins_ to setup the configuration. You'll need to install `@neo4j/graphql-plugin-auth` or a custom plugin
as shown in the xref::auth/setup.adoc[auth setup page].

> === JWT auth

Reviewer 2: Worth showing a before and after?

For JWT authorization, just pass the configuration through the `Neo4jGraphQLAuthJWTPlugin` plugin
[source, javascript, indent=0]
----
import { Neo4jGraphQL } from "@neo4j/graphql";
import { Neo4jGraphQLAuthJWTPlugin } from "@neo4j/graphql-plugin-auth";

const neoSchema = new Neo4jGraphQL({
    typeDefs,
    plugins: {
        auth: new Neo4jGraphQLAuthJWTPlugin({
            secret: "super-secret"
        })
    }
});
----

=== JWKS decoding

> https://auth0.com/docs/secure/tokens/json-web-tokens/json-web-key-sets[JSON Web Key Sets] support through `Neo4jGraphQLAuthJWKSPlugin`.

Reviewer 2: ```suggestion
https://auth0.com/docs/secure/tokens/json-web-tokens/json-web-key-sets[JSON Web Key Sets] are now supported through `Neo4jGraphQLAuthJWKSPlugin`.
```

== Types plurals changes
To improve consistency, some automatically generated plurals (e.g. `createActors`) have changed. This may cause issues if
your types use conventions such as `snake_case`.

Because of this, you may find generated queries and mutations may have different names. If you encounter this problem,
please update your clients to use the new query names or use the `plural` option in the xref::type-definitions/database-mapping.adoc#_plural[@node directive]
to force a custom plural value.

== Custom Directives
Defining and applying custom directives has changed significantly, if you are using or plan to use custom directives, make
sure to check the up-to-date documentation on xref::type-definitions/custom-directives.adoc[custom directives].

== Types changes
Some automatically generated types have changed to improve consistency.
These should not require any changes from most developers, unless types names are directly used.

Some automatically generated types have changed to improve consistency.
These should not require any changes from the developer in most cases, unless in cases where types names are directly used.

> === OnCreate Node Types

Reviewer 2: ```suggestion
=== Removal of nested operation fields for `connectOrCreate`
```

> Input types for `onCreate` in `connectOrCreate` mutations no longer accept relation fields.

Reviewer 2: ```suggestion
Input types for `onCreate` in `connectOrCreate` operations no longer accept relationship fields. They were originally added in error and did not function as one would expect, so there is no regression in functionality.
```

neoSchema.getSchema().then((schema) => {
    const server = new ApolloServer({
        schema: schema
    });
});
----

== Relationship changes
This release contains an overhaul of our relationship validations, which will require a few changes to the schema.

=== Many-to-* relationships
To improve consistency and validation, **all** "many-to-*" relationships need to be defined as _required_ in the schema:

[source, graphql, indent=0]
----
type Movie {
    actors: [Actor!]! @relationship(type: "ACTED_IN", direction: IN)
    director: Director @relationship(type: "DIRECTED", direction: IN)
}
----

Note that any other notation, such as `[Actor]` or `[Actor!]` will **not** be valid. "One-to-one" relationships
such as `Director` remain unchanged.

=== Relationship cardinality
Runtime checks for "one-to*" relationships have been added in this release, ensuring that the correct number of relationships exist. This means that some
databases with inconsistent relationships between the schema definition and the actual data may now fail in some queries.
This may have happened due to different reasons such as:
* Direct changes in the database.

> * Changes to the Type Definitions, leading to an inconsistent database.

Reviewer: More a question here but do we want this to be a list with each element on a new line? If not, ignore this comment please. If so then it isn't the case just now: https://neo4j-graphql-docs-977.surge.sh/graphql-manual/3.0/guides/v3-migration/#_relationship_cardinality

Author: Good catch. however, due to Darrell's comment this is not going to be a list, so problem solved XD

        })
    }
});
----

NOTE: Please, refer to xref::auth/setup.adoc[auth setup] before setting up auth.

== Types plurals changes
To improve consistency, some automatically generated plurals (e.g. `createActors`) have changed. This may cause issues if
your types use conventions such as `snake_case`.

Because of this, you may find generated queries and mutations may have different names. If you encounter this problem,
please update your clients to use the new query names or use the `plural` option in the xref::type-definitions/database-mapping.adoc#_plural[@node directive]
to force a custom plural value.

== Custom Directives
Defining and applying custom directives has changed significantly, if you are using or plan to use custom directives, make
sure to check the up-to-date documentation on xref::type-definitions/custom-directives.adoc[custom directives].

== Types changes
Some automatically generated types have changed to improve consistency.
These should not require any changes from most developers, unless types names are directly used.

Some automatically generated types have changed to improve consistency.

> These should not require any changes from the developer in most cases, unless in cases where types names are directly used.

Reviewer: These two blocks seem to have a lot of similar content. Shall we combine them?
